stingray/pulse/fftfit.py

```diff
@@ -0,0 +1,144 @@
+import numpy as np
+from scipy.optimize import minimize, brentq
+
+
+def find_delay_with_ccf(amp, pha):
+    nh = 32
+    nprof = nh * 2
+    CCF = np.zeros(64, dtype=np.complex)
+    CCF[:nh] = amp[:nh] * np.cos(pha[:nh]) + 1.0j * amp[:nh] * np.sin(pha[:nh])
+    CCF[nprof: nprof - nh: -1] = np.conj(CCF[nprof: nprof - nh: -1])
+    CCF[nh // 2: nh] = 0
+    CCF[nprof - nh // 2: nprof - nh: -1] = 0
+    ccf = np.fft.ifft(CCF)
+
+    imax = np.argmax(ccf.real)
+    shift = normalize_phase_0d5(imax / nprof)
+
+    return shift
+
+
+def best_phase_func(tau, amp, pha, ngood=20):
+    # tau = params['tau']
+    # good = slice(1, idx.size // 2 + 1)
+    good = slice(1, ngood + 1)
+    idx = np.arange(1, ngood + 1, dtype=int)
+    res = np.sum(idx * amp[good] * np.sin(-pha[good] + TWOPI * idx * tau))
+    # print(tau, res)
+    return res
+
+
+TWOPI  = 2 * np.pi
+
+
+def fftfit(prof, template):
+    """Align a template to a pulse profile.
+
+    Parameters
+    ----------
+    prof : array
+        The pulse profile
+    template : array, default None
+        The template of the pulse used to perform the TOA calculation. If None,
+        a simple sinusoid is used
+
+    Returns
+    -------
+    mean_amp, std_amp : floats
+        Mean and standard deviation of the amplitude
+    mean_phase, std_phase : floats
+        Mean and standard deviation of the phase
+    """
```

Comment:
I wonder if, for a long function like this, which does lots of stuff, it might be valuable to have some comments interspersed that allows the reader to broadly follow along with what's happening in the code?

Reply:
Done